File: p554m/least_bricks.py
from typing import List
from collections import defaultdict


class Solution:
    def leastBricks(self, wall: List[List[int]]) -> int:
        edge_count, max_count = defaultdict(int), 0
        for row in wall:
            s = 0
            for brick in row[0:-1]:
                s += brick
                edge_count[s] += 1
                max_count = max(max_count, edge_count[s])
        return len(wall) - max_count


# TESTS
for wall, expected in [
    ([[1, 2, 2, 1], [3, 1, 2], [1, 3, 2], [2, 4], [3, 1, 2], [1, 3, 1, 1]], 2),
    ([[3], [3], [3]], 3),
]:
    sol = Solution()
    actual = sol.leastBricks(wall)
    print("The least of crossed bricks in", wall, "->", actual)
    assert actual == expected

File: p554m/least_bricks.js
/**
 * @param {number[][]} wall
 * @return {number}
 */
var leastBricks = function (wall) {
  const edgeCount = new Map(); // <edgePos, edgeCount>
  let maxCount = 0;
  for (let row of wall) {
    let s = 0;
    for (let b = 0; b < row.length - 1; ++b) {
      s += row[b];
      let c = edgeCount.get(s) || 0;
      edgeCount.set(s, c + 1);
      maxCount = Math.max(maxCount, c + 1);
    }
  }
  return wall.length - maxCount;
};
// TESTS
[
  [
    [
      [1, 2, 2, 1],
      [3, 1, 2],
      [1, 3, 2],
      [2, 4],
      [3, 1, 2],
      [1, 3, 1, 1],
    ],
    2,
  ],
  [[[3], [3], [3]], 3],
].forEach(([wall, expected]) => {
  const actual = leastBricks(wall);
  console.log('Least number of crossed bricks ->', actual);
  console.assert(actual === expected);
});
